core/src/env3/engine/mod.rs

```diff
@@ -0,0 +1,43 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::env3::backend::{
+    Env,
+    TypedEnv,
+};
+use cfg_if::cfg_if;
+
+pub trait OnInstance: Env + TypedEnv {
+    fn on_instance<F, R>(f: F) -> R
+    where
+        F: FnOnce(&mut Self) -> R;
+}
+
+cfg_if! {
```

Comment:
That's super bad because I wanted to use this neat macro at some other places as well. Hmm. Maybe file a bug report for IntelliJ + Rust plugin? I am not sure we should guide our design by IDE bugs tbh but this sounds really weird though ...

Reply:
Seems to be a known issue: https://github.com/intellij-rust/intellij-rust/issues/4627

> I am not sure we should guide our design by IDE bugs tbh but this sounds really weird though

I agree. I can just maintain a local patch which "expands" this macro.

How do other setups handle this? I can always use VS code or vim or whatever for developing `ink!`. I at least want `go to definition` to work.

Reply:
I am using VS Code with the `rust-analyzer` extension and it works pretty well for the ink! codebase. The `go to definition` works well, too besides many other neat things. I am pretty happy with `rust-analyzer` already.